File: WikiCodeParser/Tags/LinkTag.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using WikiCodeParser.Nodes;

namespace WikiCodeParser.Tags
{
    public class LinkTag : Tag
    {
        public override string Token => "url";
        public override string Element => "a";
        public override string MainOption => "url";
        public override string[] Options => new[] {"url"};

        public override INode FormatResult(Parser parser, State state, string scope, Dictionary<string, string> options, string text)
        {
            var url = text;
            if (options.ContainsKey("url")) url = options["url"];
            if (Token == "email") url = "mailto:" + url;
            else if (!Regex.IsMatch(url, "^([a-z]{2,10}://)", RegexOptions.IgnoreCase)) url = "http://" + url;
            url = System.Web.HttpUtility.HtmlAttributeEncode(url);

            var classes = new List<string>();
            if (ElementClass != null) classes.Add(ElementClass);

            var before = $"<{Element} class=\"{String.Join(" ", classes)}\" href=\"{url}\">";
            var after = $"</{Element}>";

            var content = options.ContainsKey("url")
                ? parser.ParseTags(text, scope, IsBlock ? "block" : "inline")
                : new PlainTextNode(text);
            return new HtmlNode(before, content, after);
        }

        public override bool Validate(Dictionary<string, string> options, string text)
        {
            var url = text;
            if (options.ContainsKey("url")) url = options["url"];
            return !url.Contains("<script") && Regex.IsMatch(url, @"^([a-z]{2,10}://)?([^]""\n ]+?)$", RegexOptions.IgnoreCase);
        }
    }
}

File: php/src/Tags/LinkTag.php
<?php

namespace LogicAndTrick\WikiCodeParser\Tags;

use LogicAndTrick\WikiCodeParser\HtmlHelper;
use LogicAndTrick\WikiCodeParser\Nodes\HtmlNode;
use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\Nodes\PlainTextNode;
use LogicAndTrick\WikiCodeParser\Nodes\UnprocessablePlainTextNode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;
use LogicAndTrick\WikiCodeParser\State;

class LinkTag extends Tag
{
    public function __construct()
    {
        parent::__construct();
        $this->token = 'url';
        $this->element = 'a';
        $this->mainOption = 'url';
        $this->options = ['url'];
    }

    public function FormatResult(Parser $parser, ParseData $data, State $state, string $scope, array $options, string $text): INode|null
    {
        $url = $text;
        if (isset($options['url'])) $url = $options['url'];
        if ($this->token == 'email') $url = 'mailto:' . $url;
        else if (!preg_match('/^([a-z]{2,10}:\/\/)/i', $url)) $url = 'http://' . $url;
        $url = HtmlHelper::AttributeEncode($url);

        $classes = [];
        if ($this->elementClass != null) $classes[] = $this->elementClass;

        $element = $this->element;
        $before = "<$element " . (count($classes) > 0 ? 'class="' . implode(' ', $classes) . '" ' : '') . 'href="' . $url . '">';
        $after = "</$element>";

        $content = isset($options['url'])
            ? $parser->ParseTags($data, $text, $scope, $this->TagContext())
            : new UnprocessablePlainTextNode($text);
        return new HtmlNode($before, $content, $after);
    }

    public function Validate(array $options, string $text): bool
    {
        $url = $text;
        if (isset($options['url'])) $url = $options['url'];
        return !str_contains($url, '<script') && preg_match('/^([a-z]{2,10}:\/\/)?([^\]"\n ]+?)$/i', $url);
    }
}